### PR TITLE
Implement local blueprint support.

### PR DESCRIPTION
**Pull request by mshima**
Resolves https://github.com/jhipster/generator-jhipster/issues/19269.

---

Please make sure the below checklist is followed for Pull Requests.

- [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [ ] Tests are added where necessary
- [ ] The JDL part is updated if necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [ ] Documentation is added/updated where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by mraible** [comment]
@mshima Is there documentation that needs to be updated to show how to use this feature?

**Comment by vishal423** [comment]
I am not sure about the usefulness of this feature.  As I understand, it's trying to remove the requirement to publish a blueprint. While doing so, it's encouraging to move and maintain the blueprint code within the generated application. IMO, it defies the purpose of a blueprint to reuse it across multiple applications. Further, I don't see the npm publishing as an issue as there are ways (like `npm link`) to handle that and are already used in other maintained blueprints during development.

**Comment by vishal423** [comment]
How about identifying the pain points to improve the overall maintenance of blueprints? 

To avoid blueprint publishing, can we introduce an option like `--local <path-to-blueprint>`?

**Comment by mshima** [comment]
> How about identifying the pain points to improve the overall maintenance of blueprints?
> 
> To avoid blueprint publishing, can we introduce an option like `--local <path-to-blueprint>`?

Publishing is not required, you can install using git or path.
But you still need to:
- maintain prettier, eslint
- maintain another repository
- keep another package.json updated
- make sure the blueprint's generator-jhipster is synchronized with the project generator-jhipster
- make sure to keep the blueprint git revision updated to keep the project reproducible

Most of the tasks are minor, but makes the developer experience bad.

**Comment by vishal423** [comment]
I disagree that maintaining a blueprint in a separate repository is a bad developer experience. Infact, the opposite is a bad design choice and violates the single responsibility principle. It also increases the maintenance overhead as you need to keep each application template in sync. You can easily manage dependency updates using dependabot or equivalent tools. I don't see any challenge in that.

**Comment by mshima** [comment]
> I disagree that maintaining a blueprint in a separate repository is a bad developer experience. Infact, the opposite is a bad design choice and violates the single responsibility principle. It also increases the maintenance overhead as you need to keep each application template in sync. You can easily manage dependency updates using dependabot or equivalent tools. I don't see any challenge in that.

Yes if you want to reuse the blueprint, but in this case it's application specific.

**Comment by vishal423** [comment]
As it sounds, you want a provision to override the generated code with custom templates. Doing via current approach may be useful for a subset of specialized use cases, but it's not trivial. A blueprint is meant to be created once and reused across multiple applications. Following is what I propose:
- Create an option `--local <path-to-blueprint-template>` to allow you to specify the blueprint directory to be used to pick up and override the generated code. If you want this to be generated inside the current application, you can specify the path as `./.blueprint`.
- You can also specify the path to a full-fledged externalized blueprint as this option.
- Store the user-specified path inside `.yo-rc.json`. If the path is empty/no, then, the local blueprint option is not enabled for the current application. Generated application code need not be aware of the blueprint.
- When composed with external blueprints, a local blueprint ensures high priority.
- Create a sample repository within the `Jhipster` organization using this option with a useful override to explain/demonstrate the functional use.

**Comment by mraible** [comment]
I'm thinking of doing a Twitter poll to help answer this question. How does this sound?

> Should JHipster support local blueprints that only apply to a single app?
> 1. Yes
> 2. No, all blueprints should require publishing
> 3. Blueprints are so confusing, I'm not sure
> 4. Don't share, show answers

Please provide alternatives if my question and answers won't help.

**Comment by gmarziou** [comment]
@mraible I don't understand choice #4

**Comment by mraible** [comment]
> @mraible I don't understand choice 4

I had a typo. I meant to say "Don't care".

**Comment by gmarziou** [comment]
I see, never thought that to see survey results you have to vote. I misinterpreted the word "answers" in this context.

**Comment by mraible** [comment]
I created and published the poll. I decided to eliminate option 3 about blueprints being confusing. 

https://twitter.com/mraible/status/1563257288504123392

**Comment by mraible** [comment]
There's still 40 minutes left on my poll, but it seems like most people are in favor of local blueprints.

**Comment by deepu105** [comment]
Here is my take on this.

1. I agree that for some usecases it might be useful to have a local blueprint that adds some customization.
2. But on the other hand, given that the popularity of blueprints isn't exactly what I expected it to be, the feature might not be used a lot by regular users. So we need to look at the maintenance burden of the feature. I didn't check the code, but I expect that this will mostly use the same code as normal blueprints except gor bootstrapping from the local path
3. I'm not against this but I don't expect it to be used a lot

**Comment by mraible** [comment]
@mshima As @deepu105 mentioned, do you see this as a maintenance burden? If not, let's merge and release 7.9.3!

**Comment by mshima** [comment]
It just bootstrapping local path like @deepu105 said, I don't see as a maintenance burden.

**Comment by deepu105** [comment]
I can do the release if @DanielFran is busy

**Comment by mshima** [comment]
Can this be merged? Or should we wait for a review?

**Comment by mraible** [comment]
@deepu105 Can you please review?

**Comment by deepu105** [comment]
> @deepu105 Can you please review?

I'll take a look this week

File: generators/generate-blueprint/generator.mjs [hunk 1]
```diff
@@ -240,8 +275,8 @@ export default class extends BaseBlueprintGenerator {
 
   get postWriting() {
     return {
-      postWriting() {
-        if (this.shouldSkipFiles()) return;
+      packageJsom() {
```
Comment (deepu105) [inline generators/generate-blueprint/generator.mjs:278]: Shouldn't this be packageJson?
```suggestion
      packageJson() {
```

**Review by deepu105 — DISMISSED** [review]
I didn't nitpick. Looks good overall except for a typo

**Pull request merged**